### PR TITLE
Improve button accessibility

### LINKED ISSUE #1608: Screen reader doesn't tell when the button is pressed
We should use `aria-pressed` for that.

## Test Notes
Test with desktop & mobile (iOS & Android) screen readers
- [ ] The buttons works correctly with & without screen readers
  - [ ] The PayPal button in the payment step of the subscription dialog
  - [ ] The search button (Looking glass icon) in the search bar
  - [ ] The 'Show all features' button in the plan table on the subscription dialog
  - [ ] The about button in the settings view
  - [ ] The show attachments expander in the mail viewer
  - [ ] The 'Add folder' button in the mail view sidebar
  - [ ] The select button in the Unlock method dialog
  - [ ] The arrow buttons in the date picker
  - [ ] The arrow buttons in the sidebar calendar
  - [ ] The offline indicator in the toolbar
  - [ ] The today button in the calendar
  - [ ] The mail recipients in the mail editor
- [ ] The following other components reads correctly with screen readers:
  - [ ] The toggle buttons read their labels correctly
  - [ ] The breadcrumb in the subscription & sign up dialogs
  - [ ] The calendar checkboxes in the calendar view sidebar reads correctly
  - [ ] The calendar view selector segment control reads correctly
- [ ] The tab indexing is the same as master

### PR DESCRIPTION
**Pull request by rezbyte**
This one fixes a few things. 
Tutanota next `button-pressed-aria-1608`

- Split the `Button` component to make maintenance including accessibility enhancements easier. 
- Made the calendar page friendlier to screen readers & keyboard control.
- Made the subscriptions page friendlier to screen readers & keyboard control.
- Added keyboard control to the expanders.
- Made small styling changes to improve design consistency.

Closes #1608.

### CODE REVIEW

File: src/gui/base/OfflineIndicator.ts [hunk 1]
```diff
@@ -11,10 +10,10 @@ export const enum OfflineIndicatorState {
 }
 
 export type OfflineIndicatorAttrs =
-	| { state: OfflineIndicatorState.Online }
-	| { state: OfflineIndicatorState.Connecting }
-	| { state: OfflineIndicatorState.Synchronizing; progress: number }
-	| { state: OfflineIndicatorState.Offline; lastUpdate: Date | null; reconnectAction: () => void }
+	| { state: OfflineIndicatorState.Online; isMobile: boolean }
```
Comment (charlag) [inline src/gui/base/OfflineIndicator.ts:13]: I was thinking if we should just let it decide inside which one it is but maybe there will be cases where we want to decide it outside. It's also good when components are simple and don't decide on their own too much. Maybe we could rename it to `singleLine` or something like that?

File: src/gui/main-styles.ts [hunk 2]
```diff
@@ -300,7 +300,7 @@ styles.registerStyle("main", () => {
 			"margin-bottom": px(size.vpad_xs),
 		},
 		".mb-l": {
-			"margin-bottom": px(size.vpad_large),
+			"margin-bottom": px(size.vpad_large) + "!important",
```
Comment (charlag) [inline src/gui/main-styles.ts:303]: We should not add `!important` to these base/utility classes, it only makes sense for very specific classes like classes for specific components. If you need it in some place, please specify it inline or make a special class for your use case.

File: src/gui/main-styles.ts [hunk 3]
```diff
@@ -2146,7 +2157,7 @@ styles.registerStyle("main", () => {
 			transition: `margin-bottom 200ms ease-in-out`,
 		},
 		".circle": {
-			"border-radius": "50%",
+			"border-radius": "50% !important",
```
Comment (charlag) [inline src/gui/main-styles.ts:2160]: too me it's not obvious why `!important` is needed here

File: src/mail/view/MailFoldersView.ts [hunk 4]
```diff
@@ -156,28 +152,15 @@ export class MailFoldersView implements Component<MailFolderViewAttrs> {
...
-					color: theme.navigation_button,
-					width: `calc(100% - ${px(size.hpad_button * 2)})`,
-				},
-				onclick: () => {
-					attrs.onShowFolderAddEditDialog(attrs.mailboxDetail.mailGroup._id, null, null)
-				},
+		return m(Button, {
+			label: "addFolder_action",
+			key: "addFolder",
+			type: ButtonType.FolderRow,
```
Comment (charlag) [inline src/mail/view/MailFoldersView.ts:158]: see `ButtonType` discussion

File: src/mail/view/MailViewerHeader.ts [hunk 5]
```diff
@@ -485,37 +487,24 @@ export class MailViewerHeader implements Component<MailViewerHeaderAttrs> {
...
-											icon: BootIcons.Expand,
-											style: {
-												fill: theme.content_fg,
-												transform: this.filesExpanded ? "rotate(180deg)" : "",
-											},
-										}),
-									],
-								),
-						  ],
+						  m(ExpanderButton, {
```
Comment (charlag) [inline src/mail/view/MailViewerHeader.ts:490]: this now has a different size of the icon so it doesn't look like the other expander above it

File: src/subscription/PaymentMethodInput.ts [hunk 6]
```diff
@@ -234,24 +235,19 @@ class PaypalInput {
...
-							this.__paymentPaypalTest?.getStage(1).complete()
-							if (attrs.payPalRequestUrl.isLoaded()) {
-								window.open(attrs.payPalRequestUrl.getLoaded())
-							} else {
-								showProgressDialog("payPalRedirect_msg", attrs.payPalRequestUrl.getAsync()).then((url) => window.open(url))
-							}
-						},
+				m(Button, {
+					label: () => "PayPal",
+					type: ButtonType.PaymentLogo,
```
Comment (charlag) [inline src/subscription/PaymentMethodInput.ts:240]: see `ButtonType`  discussion

File: src/subscription/SubscriptionSelector.ts [hunk 7]
```diff
@@ -291,17 +291,12 @@ export class SubscriptionSelector implements Component<SubscriptionSelectorAttr>
...
-						onclick: (e: Event) => {
-							this.featuresExpanded[subType] = !this.featuresExpanded[subType]
-							e.preventDefault()
-						},
-					},
-					lang.get("pricing.showAllFeatures"),
-			  )
+			: m(Button, {
+					label: "pricing.showAllFeatures",
+					type: ButtonType.Expander,
```
Comment (charlag) [inline src/subscription/SubscriptionSelector.ts:296]: See `ButtonType` discussion

**Review by charlag — COMMENTED** [review]
we've tried it out and looked through the code, there are some very good accessibility changes there but I think we should reconsider some Button refactorings. Let's talk about it

**Comment by ganthern** [comment]
I think some wrapper around the attachment bubbles went missing:
image
image

**Comment by charlag** [comment]
Also in the editor

**Comment by ganthern** [comment]
anywhere where BubbleButton is used (send log / error report dialog)

**Review by ganthern — COMMENTED** [review]
Really digging this simplification. Complete list of issues that I found with this:

* BubbleButton has less margin than the usage sites of ButtonType.Bubble had previously (see above)
* toggle buttons don't change styling when toggled anymore (editor toolbar)

Banner Buttons have wrong layout:
image

**Comment by charlag** [comment]
Two things I've noticed while fixing attachment buttons:
 - pressing attachment buttons has no visual feedback. We should try to add some but this might be difficult for all cases
 - "more" menu in mobile mail header does not stop propagation and collapses the header. Oops.

**Comment by charlag** [comment]
I've changed bubble button and I've fixed in in mail editor and viewer

**Comment by rezbyte** [comment]
>  * pressing attachment buttons has no visual feedback. We should try to add some but this might be difficult for all cases

Do you mean the attachment button in the mail editor or the bubble buttons? Both seem to have feedback when I tested them but this maybe due to the commit I just pushed.

**Comment by ganthern** [comment]
> Two things I've noticed while fixing attachment buttons:
> 
>     * pressing attachment buttons has no visual feedback. We should try to add some but this might be difficult for all cases
> 
>     * "more" menu in mobile mail header does not stop propagation and collapses the header. Oops.

copy that - both things are still true. 
1. attachment buttons do not change when pressing + holding
2. the whole mail is collapsed

I also pushed a fix for the info banner layout glitch.

**Comment by rezbyte** [comment]
> copy that - both things are still true.
> 
>     1. attachment buttons and the folder column main button do not change when pressing + holding
> 
>     2. the whole mail is collapsed
> 
> 
> I also pushed a fix for the info banner layout glitch.

Thanks for the fix. The two new commits should address the other issues.

**Comment by charlag** [comment]
I would probably put the links refactoring to a separate PR unless it's also accessibility related

**Comment by rezbyte** [comment]
> I would probably put the links refactoring to a separate PR unless it's also accessibility related

It is accessibility related because the changes forces the links to be keyboard accessible. For example, the 'Website' & `Releases` links on the about dialog in settings cannot be reached by keyboard on master.

I was considering putting into another PR anyway as it depends whether a link could be considered a button or not.

**Comment by charlag** [comment]
Okay thanks for explaining, it does sound related. I think it's still better as a separate PR

**Comment by rezbyte** [comment]
> Okay thanks for explaining, it does sound related. I think it's still better as a separate PR

Alright, I will move into a new PR after lunch.

**Comment by rezbyte** [comment]
I have moved the links into [a separate PR](https://github.com/tutao/tutanota/pull/6407) now.

**Comment by charlag** [comment]
Sorry I've noticed it so late but I don't think all buttons need flash effect, especially `IconButton`s have a (newer) state layer that changes properly with interaction and opacity will mess with it a bit.

I think we should eventually replace flash effect almost everywhere (honestly it's a bit lazy implementation of interaction from old times) but for now we need it but only on certain buttons. I don't know what's the best way to do it without too much code?

**Comment by rezbyte** [comment]
> Sorry I've noticed it so late but I don't think all buttons need flash effect, especially `IconButton`s have a (newer) state layer that changes properly with interaction and opacity will mess with it a bit.
> 
> I think we should eventually replace flash effect almost everywhere (honestly it's a bit lazy implementation of interaction from old times) but for now we need it but only on certain buttons. I don't know what's the best way to do it without too much code?

Good catch, I will remove the flash from the `BaseButton` & use it per button instead..

**Comment by rezbyte** [comment]
I have re-implemented the flash using our old friend, CSS animations. The animation is now attached to just the buttons that had it originally. The original `Flash.ts` bloat has been purged.

**Review by charlag — APPROVED** [review]
Yay, it works well and  also we got rid of that weird Flash mess! Good job

**Pull request merged**